File: project8/reference/history_stacker.py
from collections import deque
from typing import Deque
import numpy as np


# Only use position and angle from the raw CartPole observation
def position_angle_from_obs(obs: np.ndarray) -> np.ndarray:
    return np.array([obs[0], obs[2]], dtype=np.float32)


class HistoryStacker:
    def __init__(self, obs_dim: int, history_len: int):
        self.history_len = history_len
        self.obs_dim = obs_dim
        self.obs_history: Deque[np.ndarray] = deque(maxlen=history_len)
        self.action_history: Deque[float] = deque(maxlen=history_len)

    def reset(self, initial_obs: np.ndarray, default_obs: float = 0.0, default_action: float = 0.0):
        self.obs_history.clear()
        self.action_history.clear()
        
        # Use default_obs value but respect obs_dim shape
        zero_obs = np.full(self.obs_dim, default_obs, dtype=np.float32)
        
        for _ in range(self.history_len - 1):
            self.obs_history.append(zero_obs)
            self.action_history.append(float(default_action))
        self.obs_history.append(initial_obs.astype(np.float32))
        self.action_history.append(float(default_action))

    def append(self, obs: np.ndarray, action: float):
        self.obs_history.append(obs.astype(np.float32))
        self.action_history.append(float(action))

    def stacked(self) -> np.ndarray:
        obs_stack = np.concatenate(list(self.obs_history), dtype=np.float32)
        act_stack = np.array(self.action_history, dtype=np.float32)
        return np.concatenate([obs_stack, act_stack], dtype=np.float32)

File: projectg_quantization8bit_onnx_linux/c_plus_plus_src/HistoryStacker.cpp
#include "HistoryStacker.h"

HistoryStacker::HistoryStacker(int obs_dim, int action_dim, int history_len)
    : obs_dim_(obs_dim), action_dim_(action_dim), history_len_(history_len) {
}

void HistoryStacker::reset(const std::vector<float>& initial_obs, float default_obs, float default_action) {
    obs_history_.clear();
    action_history_.clear();

    std::vector<float> zero_obs(obs_dim_, default_obs);
    std::vector<float> zero_action(action_dim_, default_action);

    for (int i = 0; i < history_len_ - 1; ++i) {
        obs_history_.push_back(zero_obs);
        action_history_.push_back(zero_action);
    }
    
    // Last history is the initial current observation, and zero action
    obs_history_.push_back(initial_obs);
    action_history_.push_back(zero_action);
}

void HistoryStacker::append(const std::vector<float>& obs, const std::vector<float>& action) {
    obs_history_.push_back(obs);
    if (obs_history_.size() > history_len_) {
        obs_history_.pop_front();
    }

    action_history_.push_back(action);
    if (action_history_.size() > history_len_) {
        action_history_.pop_front();
    }
}

std::vector<float> HistoryStacker::stacked() const {
    std::vector<float> result;
    // Total dim: obs_dim * history_len + action_dim * history_len
    result.reserve(obs_dim_ * history_len_ + action_dim_ * history_len_);

    // Concatenate all observations first
    for (const auto& obs : obs_history_) {
        result.insert(result.end(), obs.begin(), obs.end());
    }

    // Concatenate all actions
    for (const auto& act : action_history_) {
        result.insert(result.end(), act.begin(), act.end());
    }

    return result;
}
